File: cpp/09_bit_strings.cpp
#include <iostream>

typedef unsigned long ul;

#define MOD (ul)(1e9 + 7)

ul binPow(ul b, ul n) {
    if (n == 0) return 1;
    if (n == 1) return b % MOD;

    ul res = binPow(b, n / 2);
    res    = (res * res) % MOD;

    if ((n & 0x1) != 0) res = (res * b) % MOD;

    return res;
}

int main() {
    std::ios::sync_with_stdio(false);
    int n;
    std::cin >> n;
    ul res = binPow(2, n);

    std::cout << res << std::endl;
    return 0;
}

File: rust/src/bin/09_bit_strings.rs
use std::io;

const MOD: u64 = (1e9 + 7_f64) as u64;

fn bin_pow(b: u64, n: u64) -> u64 {
    if n == 0 {
        return 1;
    }

    if n == 1 {
        return b % MOD;
    }

    let mut res = bin_pow(b, n / 2);
    res = (res * res) % MOD;
    if (n & 0x1) != 0 {
        res = (res * b) % MOD;
    }

    res
}

fn main() {
    let mut line = String::new();
    io::stdin().read_line(&mut line).unwrap();
    let n: u64 = line.trim().parse().unwrap();
    let res = bin_pow(2, n);
    println!("{}", res);
}
